# 2022年度レポート

## レポートの諸注意

**回答方法**

- プログラムはコードセル（複数のセルでもよい）に記述する
- 問題の回答および説明はマークダウンセルに記述する

**提出方法**

1. このipynbファイルをhtml形式でエクスポートする
    - File -> Export Notebook As -> HTMLをクリック
    - 自分の分かりやすいフォルダに保存する
2. htmlファイルをポータルのオンライン授業から提出する
    - ポータルサイトのオンライン授業からレポート提出場所を開く
    - 1.で保存したhtmlファイルをアップロードする<br>
    ※ ファイル名は「学籍番号_氏名.html」とすること

**提出締切**

- 7/31（日），23:59

**評価基準**

- F評価
    - レポートに著しい不備がある／未提出
- C評価
    - レポートを期限内に提出し，基礎知識に関する問題がほとんどできている
- B評価
    - 上に加え，NumPy，Pandas，Matplotlibに関する問題の一部ができている
- A評価
    - 上に加え，NumPy，Pandas，Matplotlibに関する問題がほとんどできている
- S評価
    - 上に加え，実践編に関する課題に答えている
    - 来年度成塚のゼミを志望する場合はここまでできることが望ましい

## 基礎知識に関する問題

- 新しいマークダウンセルを下に追加し，セル内に好きな文章を記述せよ．

- 絶対パスと相対パスについて自分の言葉で説明せよ

- Windowsにおいて，コピーしたパスをそのまま貼り付けて使用すると，python側でエラーが出る．この原因と回避する方法を自分の言葉で説明せよ

- スポーツデータ分析において，Pythonを用いるメリットを自分の言葉で説明せよ

- スポーツデータ分析において，NumPy，Pandas，Matplotlibを用いるメリットを自分の言葉で説明せよ

- NumPyとPandasの違いを自分の言葉で説明せよ

- モジュールをインポートする意味を自分の言葉で説明せよ

- `os`モジュールをインポートせよ

- `os.chdir`を用いてカレントディレクトリを適当な作業フォルダに変更せよ

- NumPyを`np`，Pandasを`pd`という名前でインポートせよ

- matplotlib.pyplotを`plt`という名前でインポートせよ

- レポートを**期限内に指定された方法で**提出せよ

## NumPyに関する問題

**問題A**

以下のように，母平均5，母標準偏差0.5の正規分布に従うデータから100個を抽出した．

In [81]:
np.random.seed(seed=33)
x = np.random.normal(5, 0.5, 100)
x

array([4.841, 4.199, 4.232, 4.715, 4.892, 5.127, 4.925, 6.005, 4.952,
       5.211, 4.887, 4.681, 4.992, 5.522, 4.458, 3.897, 4.524, 5.416,
       4.5  , 5.172, 5.773, 5.345, 3.977, 5.167, 4.679, 4.889, 4.385,
       5.103, 5.411, 4.753, 4.296, 4.392, 5.867, 4.897, 4.633, 5.387,
       4.611, 4.793, 4.427, 5.163, 5.   , 5.709, 4.926, 4.697, 4.407,
       5.228, 4.581, 4.927, 4.659, 5.744, 4.453, 4.173, 4.988, 4.818,
       4.759, 5.37 , 5.306, 5.246, 5.896, 4.658, 5.527, 4.692, 5.219,
       3.595, 4.673, 5.264, 5.466, 4.944, 4.44 , 5.496, 5.815, 4.05 ,
       5.031, 5.24 , 5.19 , 4.244, 5.145, 5.048, 4.854, 5.428, 5.467,
       4.789, 5.795, 4.468, 6.052, 4.455, 5.742, 4.455, 5.302, 5.471,
       4.69 , 5.407, 4.55 , 4.76 , 5.115, 5.098, 5.268, 5.342, 6.411,
       5.458])

このデータに対し，`np.mean`関数と`np.std`関数を用いて標本平均と標本標準偏差を求めると以下のようになった．

In [82]:
np.mean(x)

4.980955414556336

In [83]:
np.std(x)

0.5156253322316987

- `np.mean`関数と`np.average`関数を使わずに`x`の標本平均を求め，上の結果と一致することを確かめよ（NumPyの他の関数は用いても良い）．
ただし，データ$x = (x_{0}, x_{1}, \ldots, x_{n})$に対して，標本平均$\bar{x}$は以下で定義される：
\begin{align}
    \bar{x} &= \frac{1}{n} \sum_{i=0}^{n} x_{i} = \frac{x_{0}+x_{1}+\cdots+x_{n}}{n}
\end{align}

- `np.std`関数と`np.var`関数を使わずに`x`の標本標準偏差を求め，上の結果と一致することを確かめよ（NumPyの他の関数は用いても良い）．
ただし，データ$x = (x_{0}, x_{1}, \ldots, x_{n})$に対して，標本標準偏差$\bar{\sigma}$は以下で定義される：
$$
    \bar{\sigma} 
    = \sqrt{\frac{1}{n} \sum_{i=0}^{n} (x_{i} - \bar{x})^2}
    = \sqrt{ \frac{(x_{0}-\bar{x})^2+(x_{1}-\bar{x})^2+\cdots+(x_{n}-\bar{x})^2}{n} }
$$

**問題B**

次のcsvファイルをダウンロードせよ：[player_England.csv](https://drive.google.com/uc?export=download&id=1C1jhTLnDg7ES3QClTf6LL34f8vXq-JgQ) <br>
このファイルには，2017年度にイングランド・プレミアリーグに所属していた選手の選手ID，身長，体重のデータが保存されている．
ただし，身長の単位はcm，体重の単位はkgである．

※ 本データはPappalardoデータセットを加工したものである（詳細は[イベントデータの解析](https://rtwqzpj5uefb1pvzmprbnq-on.drv.tw/document/講義/立正/スポーツデータ分析のためのプログラミング/6_event.html)）．

- 以下を適当に修正し，ダウンロードしたファイルをNumPy配列`D`に読み込め：

In [77]:
# csvファイルのパスを指定する
D = np.loadtxt('', delimiter=',', dtype='int')
D

array([[  3319,    180,     76],
       [    -1,   1000,   1000],
       [  3560,    179,     72],
       ...,
       [357712,    183,     92],
       [386197,    175,     71],
       [412919,    190,     76]])

配列`D`は第0列に選手ID，第1列に身長，第2列に体重が格納されている．
例えば，`D`の第0行目を見ると，選手IDが3319で身長180cm，体重76kgであることが分かる．
このデータに対し，以下の問いに答えよ．

- 選手IDが-1となっている要素はダミーデータである．`D`からダミーデータを削除し，改めて配列`D`とせよ．

- データに含まれる選手数を調べよ．

- 選手IDが703の選手の身長と体重を調べよ．<br>
※ この選手は吉田麻也選手である．2017年時点の体重と現在の体重を比較してみよ．

- 配列Dから選手ID，身長，体重のデータを抽出し，それぞれI, H, Wという配列に格納せよ．

In [62]:
I = 
H = 
W = 

- 以下の方法により，身長の最小値，最大値を求めよ
    - `H`を昇順（小→大）に並び替え，先頭と末尾の要素を抽出する
    - `np.min`，`np.max`関数を用いる

In [ ]:
# Hを昇順に並び替えて先頭と末尾の要素を抽出


In [ ]:
# np.min, np.maxを用いる


- 肥満度を表す指標としてBMIが知られている．BMIは身長と体重を用いて以下で定義される：
$$
    \mathrm{BMI} = \frac{体重 [kg]}{(身長 [m])^2}
$$
    - 身長の単位をcmからmに変換し，`H2`に格納せよ．
    - 配列`W`と`H2`からBMIを求め，`BMI`という配列に格納せよ．
    - BMIが18.5未満の選手が１人いる．この選手のIDを調べよ．<br>
    ※ この選手はRekeem Jordan Harper選手である．<br>
    ※ 日本肥満学会の基準では，BMIが18.5未満の場合を痩せ型と定義している．

In [ ]:
# Hの単位をcm -> m
H2 = 

In [ ]:
# BMIを求める
BMI = 

In [ ]:
# BMIが18.5未満を抽出


## Pandasに関する問題

次のcsvファイルをダウンロードせよ：[player_all.csv](https://drive.google.com/uc?export=download&id=1E3ahjvdekZzCu63k1oECs_GOJTS294BP) <br>
このファイルには，2017年度にヨーロッパリーグ（イングランド，フランス，ドイツ，イタリア，スペイン）に所属していた選手のデータが保存されている．<br>
※ 本データはPappalardoデータセットを加工したものである（詳細は[イベントデータの解析](https://rtwqzpj5uefb1pvzmprbnq-on.drv.tw/document/講義/立正/スポーツデータ分析のためのプログラミング/6_event.html)）．

- 以下を適当に修正し，`player_all.csv`ファイルを`df`に読み込め

In [ ]:
df = pd.read_csv('', header=0, index_col='player_id', na_values=0)
df

- `df`の先頭から2行を表示せよ

- 肥満度を表す指標としてBMIが知られている．BMIは身長と体重を用いて以下で定義される：
$$
    \mathrm{BMI} = \frac{体重 [kg]}{(身長 [m])^2}
$$
    - 身長（`height`）の単位をcmからmに変換せよ．
    - 身長（`height`）と体重（`weight`）からBMIを求め，`BMI`列を作成せよ．
    - BMIが18.5未満の選手を抽出せよ．<br>
    ※ この選手はRekeem Jordan Harper選手である．<br>
    ※ 日本肥満学会の基準では，BMIが18.5未満の場合を痩せ型と定義している．

In [ ]:
# 'height'の単位をcm->m


In [ ]:
# BMIを求めて'BMI'列を作成


In [ ]:
# BMIが18.5未満の選手を抽出


 - ポジション（`role`）ごとに，身長，体重，BMIの平均値を計算せよ．<br>

In [65]:
# ディフェンダー（'DF'）


In [66]:
# ミッドフィルダー（'MF'）


In [67]:
# フォワード（'FW'）


In [69]:
# キーパー（'GK'）


## Matplotlibに関する問題

4.で用いた[player_all.csv](https://drive.google.com/uc?export=download&id=1E3ahjvdekZzCu63k1oECs_GOJTS294BP)について，次の問いに答えよ．

**問題A：体重の箱ひげ図**
- `England`リーグに所属する選手の体重のデータから欠損値を除外したデータは以下で取得できる．
    ```python
    data1 = df.loc[df['league']=='England', 'weight'].dropna()
    ```
- 全てのリーグに対して同様のデータを求め，以下のようなリストを作成せよ
    ```python
    D = [data1, data2, ...]
    ```

- リスト`D`を用いて，体重の箱ひげ図をリーグ別に作成せよ．また，グラフを分かりやすく装飾せよ．

**問題B：身長のヒストグラム**
- `England`リーグに所属する選手の身長のヒストグラムを以下の条件で作成せよ：
    - 横軸のラベル：`height [cm]`
    - 縦軸のラベル：`Frequency`
    - 階級の数：10
    - その他の装飾は自由

## 実践編に関する課題

授業内で扱ったデータをPythonを用いて自由に解析し，解析結果をまとめよ．
- 解析内容は些細なことでも良い
- 既に知られていることでも良い